### DISCUSSION TITLE
Improve `afterSearch` event in search service

### DISCUSSION THREAD

**Discussion by gglnx** [craftcms/cms#11882] [Ideas] [answered]
Currently `EVENT_AFTER_SEARCH` of the search service only allows to track the results from the query of search index. The event doesn't include the scores or the option to modify them. I've got some possible use cases to improve this event and the search in general.

## Move `afterSearch` event before scoring

If I got some internal fields, I would like to make them searchable, but only in the control panel. If `EVENT_AFTER_SEARCH` would be triggered before the element scoring and the element scoring would use `$event->results` instead of `$results` directly, I could write a simple array filter and remove the fields.

```php
Event::on(Search::class, Search::EVENT_AFTER_SEARCH, function (SearchEvent $event) {
    if (!Craft::$app->getRequest()->getIsCpRequest()) {
        $event->results = array_filter(
            $event->results, 
            fn (array $row) => $row['fieldId'] !== '50',
        );
    }
});
```

Another use case would be to exclude fields from the search results, if the field content is only available to logged-in users or users in a special group.

This change would also allow to inject some custom results from an external API or if the client wants to add some elements manually for some special keywords.

## Add new event to modifiy scoring

A new event `afterScore` before `arsort($scoresByElementId)` would allow to modfiy the scoring. A possible use case would be to booster or to downgrade elements. This would be fixing #8995.

If wished I can provide a change request for that.

**Comment by brandonkelly** [comment, marked as answer]
Just made the following changes for Craft 4.3: (d26f4aea805ab452cae09bb7b4c8c2c57f3d20c6)

- `craft\events\SearchEvent` now has a `$scores` property.
- `craft\services\Search::searchElements()` will now trigger a `beforeScoreResults` event (`EVENT_BEFORE_SCORE_RESULTS`), and will respect any changes made to `SearchEvent::$results`. Event handlers can also set `SearchEvent::$scores` from there, which will override Craft’s internal score calculations.
- The `afterSearch` event (`EVENT_AFTER_SEARCH`) now passes the result scores to `SearchEvent::$scores`, and any changes made to it will be returned by `searchEvent()`.

**Reply by gglnx** [reply]
Perfect! Thank you ❤️

**Reply by brandonkelly** [reply]
Craft 4.3 was released earlier this week with those changes!
